src/mode/modeHandler.ts

```diff
@@ -124,10 +124,37 @@ export class ModeHandler implements vscode.Disposable {
    * Anyone who wants to change the behavior of this method should make sure
    * all selection related test cases pass. Follow this spec
    * https://gist.github.com/rebornix/d21d1cc060c009d4430d3904030bd4c1 to
-   * perform the manual testing.
+   * perform the manual testing. Besides this testing you should still test
+   * commands like 'editor.action.smartSelect.grow' and you should test moving
+   * continuously up/down or left/right with and without remapped movement keys
+   * because sometimes vscode lags behind and calls this function with information
+   * that is not up to date with our selections yet and we need to make sure we don't
+   * change our cursors to previous information (this usally is only an issue in visual
+   * mode because of our different ways of handling selections and in those cases
+   * updating our cursors with not up to date info might result in us changing our
+   * cursor start position).
    */
   public async handleSelectionChange(e: vscode.TextEditorSelectionChangeEvent): Promise<void> {
+    if (
```

Comment:
Is there any way to trigger a selection change that we should handle, but then switch editors before we get here?

Reply:
Yes. I can't exactly reproduce it all the times. But sometimes when you `Ctrl+Tab` between tabs, it will send a selection change with `e.textEditor.document` of `a.txt` when `vscode.window.activeTextEditor.document` is already at document `b.txt`, for example.
This usually ends up calling two selection change events with position (0, 0) with wrong document and then a third selection change event with the right document and the right position.
I sometimes can reproduce this when I run a debug session after closing one that already had a few files open. Then on the second session when I `Ctrl+Tab` this sometimes happens. This also happens sometimes when I first open vscode.

I've also had a situation a couple of times when this selection change event is called with `vscode.window.activeTextEditor` as `undefined`. I've not been able to reproduce this but I think this might either be some issue in vscode calling events late or it might be related to our `taskQueue`.

I decided to keep this check here because when it wasn't there it would result in a `TaskQueue: Error running task. getLineLength() called with out-of-bounds line` when we got one of this weird selection change events.

Reply:
To make it even weirder the way I found that this happens more consistently is when one of the files you're switching with is the `settings.json`. I have no idea why but this file is the one that is always making this happen to me...

Reply:
Also when this starts happening it does it with `Ctrl+Tab` and also by simply clicking on the tab to change file

Reply:
At some point soon I'd like to move away from using `window.activeTextEditor` (implicitly in our custom `TextEditor` static methods) in favor of explicitly referencing `vimState.editor`, which should eliminate those errors in this case. Once that is done we can probably remove the check, since we won't accidentally reference the wrong editor/document. For now, I'm not worried about it. This all seems like a step in the right direction, I just need to do some manual testing then I'll merge.